src/inmanta/agent/agent.py

```diff
@@ -1332,63 +1617,98 @@ async def on_disconnect(self) -> None:
         for agent_instance in self._instances.values():
             agent_instance.pause("Connection to server lost")
 
-    async def ensure_code(self, environment: uuid.UUID, version: int, resource_types: Sequence[str]) -> set[str]:
+    async def get_code(
+        self, environment: uuid.UUID, version: int, resource_types: Sequence[str]
+    ) -> tuple[list[ResourceInstallSpec], set[str]]:
         """
-        Ensure that the code for the given environment and version is loaded.
-        Return a list of all the ``resource_types`` that it failed to load.
+        Get the list of installation specifications (i.e. pip config, python package dependencies, Inmanta modules sources)
+        required to deploy a given version for the provided resource types.
         """
-        failed_to_load: set[str] = set()
         if self._loader is None:
-            return failed_to_load
+            return [], set()
 
         # store it outside the loop, but only load when required
         pip_config: Optional[PipConfig] = None
 
-        for rt in set(resource_types):
+        resource_install_specs: list[ResourceInstallSpec] = []
+        invalid_resource_types: set[str] = set()
+        for resource_type in set(resource_types):
```

Comment:
This makes the iteration order nondeterministic. Not sure if this is an issue, but if the collection has no meaningful order we should probably return `Collection[ResourceInstallSpec]` instead of `list`. I think it's good practice in general but it's especially important here to signal that the caller can not rely on order, because some sort of order is really important eventually.